### PR TITLE
Add possibility to configure graph by parameter value.

### PR DESCRIPTION
**Pull request by J-cztery**
Useful when incremental and full builds are done as part of the same job. Incremental builds do not have all compile warnings so one want to have them filter out to see the trend more clearly.

### CODE REVIEW

**Comment by uhafner** [comment]
Sorry, seems that I missed your PR. Review is coming...

File: .settings/org.eclipse.jdt.core.prefs [hunk 1]
```diff
@@ -1,396 +1,397 @@
...
+org.eclipse.jdt.core.compiler.problem.rawTypeReference=warning
+eclipse.preferences.version=1
+org.eclipse.jdt.core.compiler.problem.missingJavadocTags=warning
+org.eclipse.jdt.core.formatter.insert_space_before_closing_bracket_in_array_reference=do not insert
+org.eclipse.jdt.core.formatter.enabling_tag=@formatter\:on
+org.eclipse.jdt.core.compiler.compliance=1.6
+org.eclipse.jdt.core.formatter.blank_lines_after_package=1
+org.eclipse.jdt.core.formatter.alignment_for_superinterfaces_in_type_declaration=16
+org.eclipse.jdt.core.formatter.insert_new_line_at_end_of_file_if_missing=do not insert
+org.eclipse.jdt.core.formatter.blank_lines_before_field=0
```
Comment (uhafner) [inline .settings/org.eclipse.jdt.core.prefs]: Did you change the formatter or why is this file changed?

File: src/main/java/hudson/plugins/analysis/graph/CategoryBuildResultGraph.java [hunk 2]
```diff
@@ -234,7 +237,18 @@ protected JFreeChart createChart(final GraphConfiguration configuration, final R
         }
         return valuesPerBuild;
     }
-
+    private boolean passesFilteringByParameter(final AbstractBuild<?, ?> build, final String parameterName, final String parameterValue){
+        if (parameterName == null){
```
Comment (uhafner) [inline src/main/java/hudson/plugins/analysis/graph/CategoryBuildResultGraph.java]: Can you please check your formatter preferences? Whitespace between ( and {. Maybe it would be safe to select your chagnes and appy auto-reformat).

File: src/main/java/hudson/plugins/analysis/graph/CategoryBuildResultGraph.java [hunk 3]
```diff
@@ -234,7 +237,18 @@ protected JFreeChart createChart(final GraphConfiguration configuration, final R
...
+            return true;
+        }
+        Map<String, String> vars = build.getBuildVariables();
+        if(vars.containsKey(parameterName)){
```
Comment (uhafner) [inline src/main/java/hudson/plugins/analysis/graph/CategoryBuildResultGraph.java]: WS

File: src/main/java/hudson/plugins/analysis/graph/CategoryBuildResultGraph.java [hunk 4]
```diff
@@ -234,7 +237,18 @@ protected JFreeChart createChart(final GraphConfiguration configuration, final R
...
+            if (vars.get(parameterName).equals(parameterValue)) {
```
Comment (uhafner) [inline src/main/java/hudson/plugins/analysis/graph/CategoryBuildResultGraph.java]: Is this always non-null?

File: src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java [hunk 5]
```diff
@@ -156,15 +161,36 @@ public boolean initializeFrom(final int width, final int height) { // NOCHECKSTY
...
+     * @param dayCount
+     *            the number of days to build the graph for
+     * @param parameterName
+     *            the name of the parameter used to filter results for the graph
+     * @param parameterValue
+     *            the value of the parameter used to filter results for the graph
+     * @return <code>true</code> is the initialization was successful,
+     *         <code>false</code> otherwise
+     */
+    @SuppressWarnings("hiding")
```
Comment (uhafner) [inline src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java]: Can you please add @since tag.

File: src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java [hunk 6]
```diff
@@ -179,6 +205,26 @@ public boolean initializeFrom(final int width, final int height, final int dayCo
...
+     *            the number of days to build the graph for
+     * @param parameterName
+     *            name of parameter by which builds will be filtered for the graph
+     * @param parameterValue
+     *            value of parameter by which builds will be filtered for the graph
+     * @return <code>true</code> is the initialization was successful,
+     *         <code>false</code> otherwise
+     */
+    @SuppressWarnings("hiding")
+    public boolean initializeFrom(final String width, final String height, final String dayCountString, final String parameterName, final String parameterValue) { // NOCHECKSTYLE
```
Comment (uhafner) [inline src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java]: @since tag

File: src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java [hunk 7]
```diff
@@ -244,6 +290,10 @@ else if ("1".equals(values[5])) {
             else {
                 return false;
             }
+            if (values.length>7){
```
Comment (uhafner) [inline src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java]: ws

File: src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java [hunk 8]
```diff
@@ -317,10 +367,20 @@ private boolean initializeFromJsonObject(final JSONObject value) {
         graphType = graphId2Graph.get(grapyTypeString);
 
         useBuildDate = value.getBoolean("useBuildDateAsDomain");
+        if (value.has("parameterValue")){
+            parameterValue = value.getString("parameterValue");
+        }else{
```
Comment (uhafner) [inline src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java]: }
else {

File: src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java [hunk 9]
```diff
@@ -447,7 +509,9 @@ public String serializeToString() {
                 + buildCount + SEPARATOR
                 + dayCount + SEPARATOR
                 + graphType.getId() + SEPARATOR
-                + serializeBoolean(useBuildDate);
+                + serializeBoolean(useBuildDate)+SEPARATOR
+                + ((parameterName == null)?"":parameterName) + SEPARATOR
```
Comment (uhafner) [inline src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java]: ws

File: src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java [hunk 10]
```diff
@@ -478,20 +542,47 @@ protected String serializeBoolean(final boolean value) {
...
+     * @param newParameterValue
+     *            the parameter value
      * @return <code>true</code> if the configuration parameters are valid,
      *         <code>false</code> otherwise.
      */
     //CHECKSTYLE:OFF
     protected static boolean isValid(final int newWidth, final int newHeight,
-            final int newBuildCount, final int newDayCount, final BuildResultGraph newGraphType) {
+            final int newBuildCount, final int newDayCount, final BuildResultGraph newGraphType,
+            final String newParameterName, final String newParameterValue) {
```
Comment (uhafner) [inline src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java]: @since tag

File: src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java [hunk 11]
```diff
@@ -629,7 +737,8 @@ public boolean isVisible() {
     @Override
     public String toString() {
         return "type: " + graphType + ", size: " + width + "x" + height
-                + ", # builds " + buildCount + ", # days " + dayCount + ", useBuildDate:" + useBuildDate;
+                + ", # builds " + buildCount + ", # days " + dayCount + ", useBuildDate:" + useBuildDate
+                +",parameterName:"+parameterName+",parameterValue:"+parameterValue;
```
Comment (uhafner) [inline src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java]: ws

File: src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java [hunk 12]
```diff
@@ -433,6 +493,8 @@ private void reset() {
         dayCount = DEFAULT_DAY_COUNT;
         graphType = DEFAULT_GRAPH;
         useBuildDate = DEFAULT_USE_BUILD_DATE;
+        parameterName = null;
```
Comment (uhafner) [inline src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java]: Shouldn't it be rather the empty string?

File: src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java [hunk 13]
```diff
@@ -433,6 +493,8 @@ private void reset() {
...
+        parameterValue = null;
```
Comment (uhafner) [inline src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java]: empty string?

File: src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java [hunk 14]
```diff
@@ -478,20 +542,47 @@ protected String serializeBoolean(final boolean value) {
...
+     *            the new day count
+     * @param newGraphType
+     *            the new graph type
+     * @return <code>true</code> if the configuration parameters are valid,
+     *         <code>false</code> otherwise.
+     */
+    //CHECKSTYLE:OFF
+    protected static boolean isValid(final int newWidth, final int newHeight,
+            final int newBuildCount, final int newDayCount, final BuildResultGraph newGraphType) {
+        return isValid(newWidth, newHeight, newBuildCount, newDayCount, newGraphType, null, null);
```
Comment (uhafner) [inline src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java]: Empty String rather than null?

File: src/main/java/hudson/plugins/analysis/graph/CategoryBuildResultGraph.java (hunk 4 above)
Comment (J-cztery) [inline src/main/java/hudson/plugins/analysis/graph/CategoryBuildResultGraph.java]: "this"? Did you mean "build" ?

File: src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java (hunk 5 above)
Comment (J-cztery) [inline src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java]: Sure. What should I set it to?

File: .settings/org.eclipse.jdt.core.prefs (hunk 1 above)
Comment (J-cztery) [inline .settings/org.eclipse.jdt.core.prefs]: I don't know. Different eclipse version? What are you using?

File: src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java (hunk 12 above)
Comment (J-cztery) [inline src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java]: null better indicates no value set. Is there a specific reason for this to be empty string? Would it become empty string if this value came from GUI configuration?

File: .settings/org.eclipse.jdt.core.prefs (hunk 1 above)
Comment (uhafner) [inline .settings/org.eclipse.jdt.core.prefs]: In the meantime IntelliJ, so no worries.

File: src/main/java/hudson/plugins/analysis/graph/CategoryBuildResultGraph.java (hunk 4 above)
Comment (uhafner) [inline src/main/java/hudson/plugins/analysis/graph/CategoryBuildResultGraph.java]: No: vars.get(parameterName)
It this is null then we get a NPE. 
Can you check what happens when the parameter or value is not set in the UI. Is it the empty string or null?

File: src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java (hunk 5 above)
Comment (uhafner) [inline src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java]: The next version, just remove the snapshot from the pom version.

File: src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java (hunk 12 above)
Comment (uhafner) [inline src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java]: See comment above, is empty string default in UI or null?

File: src/main/java/hudson/plugins/analysis/graph/CategoryBuildResultGraph.java (hunk 4 above)
Comment (J-cztery) [inline src/main/java/hudson/plugins/analysis/graph/CategoryBuildResultGraph.java]: Sorry, i saw it on different view and did not realize you pointed out the specific line. You are right there are some potential NPE's here. I will add checks.

File: src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java (hunk 12 above)
Comment (J-cztery) [inline src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java]: When no parameter and value is set and it is serialized to defaults file it will take form: 500!200!50!30!PRIORITY!0!!! . 
Now if you feed a string like this to StringUtils, i.e : 
String[] values = StringUtils.split(value, SEPARATOR);
It will have only 6 values. So there is no way to distinguish between empty string and null in defaults file. 
So i am not sure what you want me to do here ;)

**Comment by J-cztery** [comment]
Huh. Had a small nightmare with git but it should be more all less all right. What else do you thing is outstanding?

File: src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java (hunk 12 above)
Comment (uhafner) [inline src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java]: I don't understand this comment. 

I generally think it is safer to use empty strings rather than null values to avoid NPE. If this is not possible, then using null is ok.

File: src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java [hunk 15]
```diff
@@ -447,7 +513,9 @@ public String serializeToString() {
                 + buildCount + SEPARATOR
                 + dayCount + SEPARATOR
                 + graphType.getId() + SEPARATOR
-                + serializeBoolean(useBuildDate);
+                + serializeBoolean(useBuildDate) + SEPARATOR
+                + ((parameterName == null) ? "" : parameterName) + SEPARATOR
```
Comment (uhafner) [inline src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java]: This would be simpler with empty string.

**Comment by uhafner** [comment]
There are still a lot of formatting problems but I think I can auto-format that part later. I need to activate CheckStyle in our CI build so we would see those problems easier. (Do you have the checkstyle plug-in installed in Eclipse? Then you should also see these warnings.) 

In general, I prefer the final PR with just one commit, can you please squash the commits to a single one?

**Comment by J-cztery** [comment]
Easy. Can i do this without creating new branch? Is it ok to create new branch and pull request?
I will tell you more about Eclipse later today, when i am home.

**Comment by uhafner** [comment]
Yes: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History
This modifies the current branch.

The first time you do that it might be somewhat complicated... But then you know how to proceed for the next times;-)

**Comment by J-cztery** [comment]
Looks like my ordeal is coming to an end ;) That was a very good link. Thanks!!! 
Ok, what i did was: 
 mvn -DdownloadSources=true -DdownloadJavadocs=true -DoutputDirectory=target/eclipse-classes eclipse:eclipse
to generate eclipse project. 
I am using Eclipse Version: 4.2.0. I do not have checkstyle plugin installed. 
I can see your templates but when i asked eclipse to reformat file it reformatted mostly documentation ;)

Sometimes(for various plugins) project in eclipse does not even compile but it is enough to step through the code. My natural environment is Visual Studio so i am not really eclipse or java fluent. 
Anyway. Thank you!

File: src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java (hunk 12 above)
Comment (uhafner) [inline src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java]: Did you see my comment above?

**Comment by uhafner** [comment]
mvn eclipse:eclipse is outdated and should not be used anymore. m2eclipse should do a better job.

**Comment by uhafner** [comment]
Can you please add a test to GraphConfigurationTest? testInvalidConfiguations needs additional values, too.

File: src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java (hunk 12 above)
Comment (J-cztery) [inline src/main/java/hudson/plugins/analysis/graph/GraphConfiguration.java]: I said that as soon as you serialize PARAMETER_VALUE  and PARAMETER_NAME to file of cookie the difference disappears.
I do not see any benefits of changing this code but feel free to change it as you see fit.

**Comment by J-cztery** [comment]
I will try m2eclipse and adding an unit test

**Comment by J-cztery** [comment]
Ok, added parameters to unit test. And found a problem in equals! ;)

**Comment by uhafner** [comment]
Thanks for your contribution. I changed it slightly, hopefully everything still works in your setup :-)
